content/chapters/data/lab/content/memory-security.md

```diff
@@ -1,42 +1,100 @@
 # Memory Security
 
+Memory security is one of the most important aspects in today's computer systems.
+Its main objective is to avoid the development of vulnerable code and prevent attackers from exploiting existing memory vulnerabilities.
+Memory protection techniques are implemented at all levels of memory abstraction: hardware, operating system and programming language.
+In addition, third-party software tools may be used to statically and dynamically examine a program to identify vulnerabilities when working with memory.
+
+In this section, we will focus on the main memory vulnerabilities and how to exploit them.
+We will use the C programming language for presentation purposes, however, these examples may be reproduced in any language that implements arrays as pointers without bounds and allows the liberal use of pointers.
+
+## Wild Pointer Arithmetic Info Leak
+
+In C, once a pointer has been initialized with a valid address it can potentially access any location of the process address space.
+This is problematic for situations where the binary contains sensitive information.
+Take for example a server that authenticates users: if the password is stored in some buffer, then if a pointer is wrongfully used, it can end up leaking the password.
+
+Navigate to the `support/memory-security/` directory.
+Open and analyze the `buff-leak.c` file.
+
+[Answer this quiz](../quiz/memory-regions-vars.md)
+
+### Practice
+
+The pointer `p` points to the stack, however, we can modify any variable that is declared in the program through `p`.
+All we need to know is the offset of the other memory locations that we wish to access.
+Run the program and try to input the correct offsets to modify variables from different regions of our program.
+Once a correct offset is given as input, the program will output a validation message.
+
+Note that adding or subtracting user provided values to pointers enables an attacker to observe a programs entire memory!
+
 ## Buffer Overflow Info Leak
 
-TODO
+Since arrays decay to pointers in C, the same effect may be obtained by using them.
 
-### Quiz
+Navigate to the `support/memory-security/` directory.
+Open and analyze the `array-leak.c` file.
 
-TODO
+Compile and run the program.
+Can you extract any information from the output?
+Can you identify the return address of the main function?
+
+**Note:** You can use `objdump -D -M intel array_leak` to check the stack layout.
```

Comment:
```suggestion
**Note:** You can use `objdump -d -M intel array_leak` to check the stack layout.
```
Disassembling the entire executable isn't necessary and adds unnecessary output. Simply disassembling the code sections should suffice.